### ISSUE TITLE
Vmap Over Haiku Params

### ISSUE THREAD

**Issue by stolet** [deepmind/dm-haiku#219]
I have a list of haiku parameters for a neural network and I want to use my network to make a forward pass and compute a gradient update for each set of parameters in the list. The naive way to do this would be to loop over each parameter in the list and compute the gradient. This is a bit slow, so I thought this was a good opportunity to use the vmap function in jax, where the vmap maps the grad of the loss function to each each param.

For example, I want to have something like this

```
def loss_fn(param, inp, target):
  pred = model.apply(param, inp)
  loss = cross_entropy_loss(pred, target)
  return loss

vmap_grads = jax.vmap(jax.value_and_grad(loss_fn), in_axes=(0, 0, 0))
loss, grads = vmap_grads(params, inp, target)
```

Where `params` is a list of params for the model and it gets vmapped so that each parameter in this list of parameters gets mapped to the loss function. This doesn't work because params is a regular python list where each element in the list is a pytree with the parameters for each layer. Is there a way to do this in an efficient way with vmap or am I stuck having to do this in a slow for loop?

**Comment by LenaMartens** [comment]
Hi! You can definitely `vmap` over a Haiku apply function, but you're right that having a list of `params` is the wrong format to do this with. You want to stack your parameter arrays and end up with a `params` dict where every `param` array has a dimension which will become the `vmapped` dimension.

For example, if you have three `param` dictionaries you want to `vmap` over, you could stack them like:

```
params = [params1, params2, params3]
stacked_params = jax.tree_map(lambda *arrays: jnp.stack(arrays), *params)
print(jax.tree_map(lambda x: x.shape, stacked_params))
>> FlatMap({
  'mlp/~/linear_0': FlatMap({'b': (3, 20), 'w': (3, 100, 20)}),
  'mlp/~/linear_1': FlatMap({'b': (3, 10), 'w': (3, 20, 10)}),
})
```

Does that make sense?

**Comment by stolet** [comment]
That makes sense. Thanks!
